sfs/time/drivingfunction.py

```diff
@@ -255,3 +256,205 @@ def apply_delays(signal, delays):
+        Speed of sound
+
+    Returns
+    -------
+    delay : float
+        Overall delay in seconds (common to all secondary sources)
+    weight : float
+        Overall weight (common to all secondary sources)
+    sos : list of array_like
+        Second-order section filters
+    phaseshift : float
+        Phase shift in radians
+
+    References
+    ----------
+        S. Spors, V. Kuscher, J. Ahrens (2011) - "Efficient realization of
+       model-based rendering for 2.5-dimensional near-field compensated higher
+       order Ambisonics", WASPAA, p. 61-64
+
+       See Eq.(11)
+
+    """
+    max_order = util.max_order_circular_harmonics(len(x0), max_order)
+    if c is None:
+        c = defs.c
+
+    x0 = util.asarray_of_rows(x0)
+    xs = util.asarray_1d(xs)
+    phi0, _, _ = util.cart2sph(*x0.T)
+    phi, _, r = util.cart2sph(*xs)
+
+    delay = (r-r0)/c
+    weight = 1/2/np.pi/r
+    sos = []
+    for m in range(max_order+1):
+        _, p, k = besselap(m, norm='delay')
+        # TODO: modify "besselap" for very high orders (>150)
+        s0 = (c/r)*p
+        sinf = (c/r0)*p
+        z0 = np.exp(s0/fs)
+        zinf = np.exp(sinf/fs)
+        # TODO: select matched-z or bilinear transform
+        k = _normalize_digital_filter_gain(s0, sinf, z0, zinf, fs)
+        sos.append(zpk2sos(z0, zinf, k, pairing='nearest'))
+        # TODO: normalize the SOS filters individually?
+    phaseshift = phi0 - phi
+    return delay, weight, sos, phaseshift
+
+
+def nfchoa_driving_signals(delay, weight, sos, phaseshift, signal, max_order=None):
+    """Get NFC-HOA driving signals per secondary source.
+
+    Parameters
+    ----------
+    delay : float
+        Overall delay in seconds (common to all secondary source)
+    weight : float
+        Overall weight (common to all secondary sources)
+    sos : list of array_like
+        Second-order section filters
+    phaseshift : (C,) array_like
+        Phase shift in radians
+    signal : tuple of (N,) array_like, followed by 1 or 2 scalars
+        Excitation signal consisting of (mono) audio data, sampling rate
+        (in Hertz) and optional starting time (in seconds).
+    fs: int
+        Sampling frequency in Hz
+    max_order : int, optional
+
+    Returns
+    -------
+    driving_signals : (N, C) numpy.ndarray
+        Driving signals.
+    t_offset : float
+        Simulation point in time offset (seconds).
+
+    """
+    max_order = util.max_order_circular_harmonics(len(phaseshift), max_order)
+    data, fs, t_offset = util.as_delayed_signal(signal)
+    N = len(phaseshift)
+    L = len(data)
+    d = np.zeros((L, N))
+#    TODO : check FOS/SOS structure
+
+    d = np.tile(np.expand_dims(sosfilt(sos[0], data), 1), (1, N))
+    for m in range(1, max_order+1):
+        modal_response = sosfilt(sos[np.abs(m)], data)
+        for n in range(N):
+            d[:, n] += modal_response * 2 * np.cos(m*phaseshift[n])
+    t_offset += delay
+    return np.real(d) * weight, fs, t_offset
```

Comment:
This should return a `DelayedSignal`, right?

And why is `np.real()` used here?

Reply:
This is also something I need before when the `modal_response` was multiplied with a complex term.